litestar/app.py

```diff
@@ -450,6 +451,33 @@ def __init__(
             config.stores if isinstance(config.stores, StoreRegistry) else StoreRegistry(config.stores)
         )
 
+    @property
+    def cli_plugins(self) -> list[CLIPluginProtocol]:
+        return self.get_plugins_of_type(CLIPluginProtocol)
```

Comment:
I have no idea why the typing doesn't work here.. I feel like I made some very obvious mistake that I'm missing completely..

Reply:
I don't have a complete and coherent explanation either but I think it comes down to the annotation of `self._plugins`. `Mapping[type[PluginProtocol], PluginProtocol]` is not strict enough based on the intended usage: it allows mapping a type `A` to an instance of a different type `B` as long as both `A` and `B` are PluginProtocols. I don't know if the desired stricter annotation is currently expressible.

Other observations:
- The properties give a different error based on the presence or absence of `lru_cache`:
   - With `lru_cache`: `Incompatible return value type (got "list[PluginT]", expected "list[SerializationPluginProtocol]")  [return-value]`
   - Without `lru_cache`: `Only concrete class can be given where "type[SerializationPluginProtocol]" is expected  [type-abstract]`
- Changing the signature of `get_plugins_of_type` from `(self, type_: type[PluginT]) -> list[PluginT]` to `(self, type_: type[PluginProtocol]) -> list[PluginProtocol]` is more consistent with the `self._plugins` annotation. In this case the error is the same with and without `lru_cache`: `Incompatible return value type (got "list[SerializationPluginProtocol | InitPluginProtocol | OpenAPISchemaPluginProtocol | CLIPluginProtocol]", expected "list[SerializationPluginProtocol]")  [return-value]` 

As for what to do about it I can think of two options:
1)  
![7u8dh4](https://github.com/litestar-org/litestar/assets/291289/11b17e77-11f4-43a9-862f-5ae488c1dfc5)

2) Abandon the generic attempt and inline `get_plugins_of_type` on every property:
```py
    @cached_property
    def cli_plugins(self) -> list[CLIPluginProtocol]:
        return [p for p in self._plugins.values() if isinstance(p, CLIPluginProtocol)]

    @cached_property
    def openapi_schema_plugins(self) -> list[OpenAPISchemaPluginProtocol]:
        return [p for p in self._plugins.values() if isinstance(p, OpenAPISchemaPluginProtocol)]

    @property
    def serialization_plugins(self) -> list[SerializationPluginProtocol]:
        return [p for p in self._plugins.values() if isinstance(p, SerializationPluginProtocol)]
```

Reply:
Thanks for the input @gsakkis!

So, I did some more digging and found something curious. Removing `@lru_cache` reveals a different error:

```python
    @property
    def cli_plugins(self) -> list[CLIPluginProtocol]:
        return self.get_plugins_of_type(CLIPluginProtocol) #  error: Only concrete class can be given where "type[CLIPluginProtocol]" is expected  [type-abstract]
```

I'm now wondering whether this is an issue with mypy, `@lru_cache`'s typing, or my code..
Probably a mix? :upside_down_face: 

Interestingly enough, pyright doesn't report issues anymore without the `@lru_cache`.

<hr>

Regarding your suggestions: They would work, but kinda defeat the purpose. The idea I had was that users could use this function to retrieve these plugins in a type safe manner. The properties are just for backwards compatibility.

If it turns out this simply can't be typed properly (which I'm not yet sure, since it *looks* like this should work!), I'd rather just remove the method in favour of exposing the `plugins` mapping directly, so users can iterate over it themselves (and get correct typing).